### DISCUSSION TITLE
Provide pre-build docker images

### DISCUSSION THREAD

**Discussion by JJ-8** [TFNS/CTFNote#211] [Ideas]
I get quite often the comment that CTFNote requires a large amount of memory to host (see also https://github.com/TFNS/CTFNote/issues/126). This is also my own experience, since it requires around 2-3 GB RAM to build the frontend. However, afterward the building of especially the frontend, CTFNote requires much less RAM to run. 

I think it would be nice to provide prebuilt docker images hosted via Github Container Registry or a similar service (which one?). This way, you can bypass the building and host CTFNote even on a 2 GB RAM VPS.
quay.io
However, I think we should not drop the manual build procedure since it may be preferred to not trust the prebuilt images and an easy alternative should be provided. But we may change the default docker-compose.yml file to use the prebuilt images.

Would this be supported by the CTFNote maintainers?

**Comment by XeR** [comment]
> I think it would be nice to provide prebuilt docker images hosted via Github Container Registry or a similar service (which one?). This way, you can bypass the building and host CTFNote even on a 2 GB RAM VPS.

I agree.

I'd love to get more familiar with Github Actions (or similar) and have actions that automatically build and deploy packages and containers.
I think we should have a container tag for each git tag (e.g. have a new container for each release) and use the `latest` tag for `HEAD`.
Unfortunately, I have never used Github Actions so I have no idea how it works. I don't want to mess up and leak secrets.
@anthares101's [fork], which is the most popular image on [Docker Hub] is probably a good place to start.

[fork]: https://github.com/anthares101/CTFNote/blob/main/.github/workflows/publish.yaml
[Docker Hub]: https://hub.docker.com/r/anthares101/ctfnote

> via Github Container Registry or a similar service (which one?)

[GHCR] free plan has a 500MB limit.
[Quay] gives away unlimited storage space for FOSS projects.
[Hub] has a limit of 100 pulls every 6 hours

[GHCR]: https://github.com/features/packages#pricing
[Quay]: https://quay.io/plans/
[Hub]: https://www.docker.com/pricing/

I feel like we would reach the 500MB relatively fast.
```
% docker image list ctfnote-api
REPOSITORY    TAG       IMAGE ID       CREATED        SIZE
ctfnote-api   latest    68b6233fcb76   26 hours ago   313MB
```

The 100 pulls limit on the other hand seems very high. (How often do *you* deploy a new version of CTFNote ?)
Do people even care about the containers used ? e.g. corporate policy that requires the use of Docker Hub.
If not, quay is the clear winner here. It seems to pull repositories, compile them and publish them automatically.
See this [Gitlab blog post](https://about.gitlab.com/blog/2015/08/19/quayio/).

> However, afterward the building of especially the frontend, CTFNote requires much less RAM to run.

If you are low on memory, here is how I compile the front for my server :

```diff
diff --git a/front/Dockerfile b/front/Dockerfile
index 4c35f732..19911867 100644
--- a/front/Dockerfile
+++ b/front/Dockerfile
@@ -11,6 +11,10 @@ WORKDIR /usr/src/app
 
 RUN yarn build
 
+# _--------_
+FROM scratch AS export
+COPY --from=build-deps /usr/src/app/dist/spa /
+
 # _--------_
 FROM nginx@sha256:5ba534070ae1e5e83d52141b11ddced689b476c0001e7205f50979dc0cbdde3d
```

Then I compile with `docker image build --target export --output - . > www.tar` on my own machine.
I upload the archive on my server and extract it. Since I already have a nginx server, I configured it and do not run the `front` container saving additional MB of memory.

Questions:
1. How does one learn Github Actions ?
2. Can I run a Github Action server locally, or is it 100% proprietary ?
3. Do we even *want* to use Github Action ? Quay looks dead simple, and I like simple.

**Reply by JJ-8** [reply]
Cool to see your fix for this issue! Thanks for sharing :)

In my fork of CTFNote I already fixed this problem by implementing a workflow to build and publish docker images to GHCR: https://github.com/JJ-8/CTFNote/tree/0-pre-build-docker-images-through-ghcr. Check the README for the updated instructions since I created a hybrid mode: you can build the container yourself but you can also use the prebuild one. This is done through a `docker-compose.prebuild.yml` file which extends the original `docker-compose.yml`.

> Do we even want to use Github Action ? Quay looks dead simple, and I like simple.

Choosing the hosting provider does not make it easier or more difficult. It just has different restrictions. The workflow I have created uses [docker login action](https://github.com/docker/login-action) which supports GHCR, Docker Hub and Quay.io. If you configure the repository secrets correctly, it is easy to support all of them. However, since GHCR is integrated with GitHub and requires the least configuration, I have used that one currently.

>  I think we should have a container tag for each git tag (e.g. have a new container for each release) and use the latest tag for HEAD.

This is some logic that is definitely possible with the docker actions, but are a bit more complicated to realize. If we want to keep it simple, the easiest solution is to publish a docker image to `latest` on every new commit to the main branch (my current implementation).

The fork you linked looks quite interesting to combine with my fork. For example, the building of an arm image is a good addition.

If we want to support only GHCR for now, it is very easy for me to create a PR with the code I have now. However, if we want to support more/other targets, someone should go through creating an account, configuring the secrets and adding it to the docker login action. That requires a bit more work and specific privileges.

**Reply by Anthares101** [reply]
If you end up wanting to combine forks, I will be glad to help if you need it 😄 . My action is not really managing Docker tags (Just latest tag used) because to be honest did not expect the popularity of the images, I just needed a way of deploying to my K3S ARM cluster and I built something fast, but should be easy to implement that management.

About ARM, I use QEMU for cross-compiling the image in a AMD64 environment (No native ARM for now in actions 😢) it obviously take some time but using the cache is no too bad.

I use the same login action for DockerHub setup, I guess with GHCR you avoid the "DockerHub site configuration" but the free limit is pretty tight.

**Reply by JJ-8** [reply]
@anthares101, thanks for joining the discussion!

> My action is not really managing Docker tags (Just latest tag used) because to be honest did not expect the popularity of the images, I just needed a way of deploying to my K3S ARM cluster and I built something fast, but should be easy to implement that management.

Do you have an example of how to implement this? The only way I know is through the [docker/metadata-action](https://github.com/docker/metadata-action).

> I use the same login action for DockerHub setup, I guess with GHCR you avoid the "DockerHub site configuration" but the free limit is pretty tight.

GHCR just works, but is not the best option to choose I think. You always need a Github personal access token in order to pull the images, which takes time to setup. Something like Docker Hub would be better since you can access that anonymously.

**Reply by Anthares101** [reply]
> Do you have an example of how to implement this? The only way I know is through the [docker/metadata-action](https://github.com/docker/metadata-action).

Yeah same, maybe there is another option but I don't really know about it. The examples in the [docker/metadata-action](https://github.com/docker/metadata-action) readme seems pretty solid, maybe some modifications needed but not much.

**Reply by JJ-8** [reply]
I have updated the workflow in my fork, tested it and created a PR: https://github.com/TFNS/CTFNote/pull/217. Feel free to leave any comments and/or suggest improvements! Scope of the PR is to only add GHCR for now and as soon as we have decided which Docker platform to use, add that one as well in a future PR.

**Reply by Anthares101** [reply]
The PR looks great, did not know about the matrix thing for jobs paralel execution 😲. Thanks for including the ARM build ❤️.

**Comment by JJ-8** [comment]
For some time we already provide images through GHCR which works solid, so we can close this discussion.
